Python/Product/TestAdapter/TestContainerDiscovererProject.cs

```diff
@@ -328,81 +331,101 @@ class TestContainerDiscovererProject : ITestContainerDiscoverer, IDisposable {
                 return;
 
             if (IsSettingsFile(e.File)) {
-                switch (e.ChangedReason) {
-                    case TestFileChangedReason.None:
-                        break;
-                    case TestFileChangedReason.Renamed:
-                        _testFilesUpdateWatcher.RemoveWatch(e.OldFile);
-                        _testFilesUpdateWatcher.AddWatch(e.File);
-                        break;
-                    case TestFileChangedReason.Added:
-                        _testFilesUpdateWatcher.AddWatch(e.File);
-                        break;
-                    case TestFileChangedReason.Changed:
-                        _testFilesUpdateWatcher.AddWatch(e.File);
-                        break;
-                    case TestFileChangedReason.Removed:
-                        _testFilesUpdateWatcher.RemoveWatch(e.File);
-                        break;
-                }
+                HandleSettingsChanged(e);
+            }
+            // VS uses temp files we need to ignore
+            else if (IsTestFile(e.File)) {
+                HandleSourcesChanged(e);
+            }
+            return;
+        }
 
-                _isRefresh = true;
-                NotifyContainerChanged();
+        private void HandleSourcesChanged(TestFileChangedEventArgs e) {
+            var projectInfo = FindProjectInfo(e.File, e.Project);
+            if (projectInfo == null)
                 return;
+
+            var sources = new List<string>() { e.File };
+            switch (e.ChangedReason) {
+                case TestFileChangedReason.Added:
+                    UpdateContainersAndListeners(sources, projectInfo, isAdd: true);
+                    break;
+                case TestFileChangedReason.Changed:
+                    //Need to increment version number so Test Explorer notices a change
+                    UpdateContainersAndListeners(sources, projectInfo, isAdd: true);
+                    break;
+                case TestFileChangedReason.Removed:
+                    UpdateContainersAndListeners(sources, projectInfo, isAdd: false);
+                    break;
+                case TestFileChangedReason.Renamed:
+                    var oldFileProjectInfo = FindProjectInfo(e.OldFile);
+                    UpdateContainersAndListeners(new List<string>() { e.OldFile }, oldFileProjectInfo, isAdd: false);
+                    UpdateContainersAndListeners(sources, projectInfo, isAdd: true);
+                    break;
+                default:
+                    //In changed case file watcher observed a file changed event
+                    //In this case we just have to fire TestContainerChnaged event
+                    break;
             }
 
-            // VS uses temp files we need to ignore
-            if (!IsTestFile(e.File))
-                return;
+            if (ShouldRebuild(e.File) || ShouldRebuild(e.OldFile)) {
+                _isRefresh = true;
+            }
```

Comment:
this is the fix for when adding  __init__.py

Its a bit overkill, I was going to just grab the files in the same directory, but then I should also check child folders.

Alternatively I could try unloading and reloading the project or manually looping through the project to update the timestamp on the testscontainers.